File: src/main/kotlin/ca/yakov/aoc/aoc2016/day17.kt
package ca.yakov.aoc.aoc2016

import ca.yakov.aoc.Solution
import com.soywiz.krypto.md5

private data class Event(val path: String, val x: Int, val y: Int) {
    fun up() = Event(path + "U", x, y - 1)
    fun down() = Event(path + "D", x, y + 1)
    fun left() = Event(path + "L", x - 1, y)
    fun right() = Event(path + "R", x + 1, y)
}

fun run17(inputUntrimmed: String): Solution {
    val input = inputUntrimmed.trim()
    val events = mutableListOf(Event("", 0, 0))
    var shortest = ""
    var longest = 0uL

    while (events.isNotEmpty()) {
        val event = events.removeFirst()
        if (event.x == 3 && event.y == 3) {
            if (shortest.isEmpty()) {
                shortest = event.path
            }
            longest = event.path.length.toULong()
            continue
        }

        val hash = "$input${event.path}".encodeToByteArray().md5()
        if (event.x < 3 && (hash.bytes[1].toInt() and 0xf) >= 0xb) {
            events += event.right()
        }
        if (event.y < 3 && (hash.bytes[0].toInt() and 0xf) >= 0xb) {
            events += event.down()
        }
        if (event.x > 0 && (hash.bytes[1].toInt() and 0xff) >= 0xb0) {
            events += event.left()
        }
        if (event.y > 0 && (hash.bytes[0].toInt() and 0xff) >= 0xb0) {
            events += event.up()
        }
    }

    return Solution(p1 = 0uL, s1 = shortest, p2 = longest)
}

File: src/main/nim/aoc/2016/day17.nim
import std/md5, ../aoc

type
  Event = object
    path: string
    x: int
    y: int

proc up(event: Event): Event = Event(path: event.path & 'U', x: event.x, y: event.y - 1)
proc down(event: Event): Event = Event(path: event.path & 'D', x: event.x, y: event.y + 1)
proc left(event: Event): Event = Event(path: event.path & 'L', x: event.x - 1, y: event.y)
proc right(event: Event): Event = Event(path: event.path & 'R', x: event.x + 1, y: event.y)

proc isDoorOpen(c: char): bool = int(c) >= int('b')

proc day17*(file: string): Solution =
  var events = @[Event(path: "", x: 0, y: 0)]
  var shortest = ""
  var longest = 0

  while events.len() != 0:
    let event = events[0]
    events.delete(0)
    if event.x == 3 and event.y == 3:
      if shortest.len() == 0:
        shortest = event.path
      longest = event.path.len()
      continue

    let hash = (file & event.path).getMD5()
    if event.y > 0 and hash[0].isDoorOpen(): events &= event.up()
    if event.y < 3 and hash[1].isDoorOpen(): events &= event.down()
    if event.x > 0 and hash[2].isDoorOpen(): events &= event.left()
    if event.x < 3 and hash[3].isDoorOpen(): events &= event.right()

  Solution(s1: shortest, s2: $longest)
